### ISSUE TITLE
go http redirect caveats

### ISSUE THREAD

**Issue by mgood** [ajbouh/substrate#166]
I noticed that requests for `/bridge2` are now redirecting to `/sessions` rather than `/bridge2/sessions`. It was updated to use `SUBSTRATE_URL_PREFIX` to detect the "base URL" for things like redirects, but it seems like that may not be working anymore.

**Comment by mgood** [comment]
This seems to be a combination of two things:
In #163 I changed the URLs slightly, so instead of `/sessions` it would be `/sessions/new` to start a new session. So `/sessions/` (with trailing slash) would show just the "index". However, on a request for `/sessions` Go would implicitly redirect to add the trailing `/`, but it's unaware of the URL prefix.
For that, Go sends a "permanent redirect", which the browser then cached, so even on `main` the browser caching would try to redirect `/bridge2(...)/sessions` to `/sessions`.

So, I can update the new bridge2 code to redirect `/` to `/sessions/` with a trailing slash.
However, that implicit trailing slash redirect could still bite us in the future, so it might be helpful to figure out how to override that in such a way that it's either disabled, or adds the prefix.

**Comment by ajbouh** [comment]
Why does it add the trailing slash automatically?

**Comment by mgood** [comment]
It seems to be a feature for convenience in Go's `ServeMux`. If you go to `/foo` which doesn't have a handler, but `/foo/` does, it assumes that the canonical path should have a trailing slash and redirects them.

**Comment by ajbouh** [comment]
Got it, thanks for explaining all that and catching it in the first place. This can go into #163 before we merge, right? That would explain why I hadn't seen the problem.

**Comment by mgood** [comment]
Yeah, I updated #163 to solve the immediate issue, though this is seeming like an annoyingly easy trap to fall into.
My first quick change to add the trailing slash like `path.Join(basePath, "sessions/")` failed because `path.Join` was calling `Clean` which removed that trailing slash, triggering the same redirect issue.

I started writing a "middleware" to rewrite the `Location` header in the response to add the prefix. That might be reasonable, but it started adding complexity since just wrapping `ResponseWriter` then doesn't expose other interfaces like `http.Hijacker` which is used by the WebSocket implementation to hijack the connection. That could be fixed, but I figured it was better at this point to hold off on a more general solution.

It also seems like these redirects are easily broken if you use things like the `StripPrefix` middleware. It simply changes the request path, which would then break those redirects since they don't see the full path to append the `/` to.
